pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -552,26 +640,42 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListenerForPortOrUDS(no
 	}
 
 	var allChains []plugin.FilterChain
-
 	for _, p := range configgen.Plugins {
 		chains := p.OnInboundFilterChains(pluginParams)
 		allChains = append(allChains, chains...)
 	}
-	// Construct the default filter chain.
-	if len(allChains) != 0 {
-		log.Debugf("Multiple plugins setup inbound filter chains for listener %s, FilterChainMatch may not work as intended!",
-			listenerMapKey)
-	} else {
+
+	if len(allChains) == 0 {
 		// add one empty entry to the list so we generate a default listener below
 		allChains = []plugin.FilterChain{{}}
 	}
 
+	tlsInspectorEnabled := false
+allChainsLabel:
+	for _, c := range allChains {
+		for _, lf := range c.ListenerFilters {
+			if lf.Name == wellknown.TlsInspector {
+				tlsInspectorEnabled = true
+				break allChainsLabel
+			}
+		}
+	}
+
+	var filterChainMatchOption []FilterChainMatchOptions
 	// Detect protocol by sniffing and double the filter chain
 	if pluginParams.ListenerProtocol == plugin.ListenerProtocolAuto {
 		allChains = append(allChains, allChains...)
+		if tlsInspectorEnabled {
```

Comment:
I'm extremely confused by this. Why is the decision to install PERMISSIVE or STRICT filter chains based on whether or not the TLS Inspector was already configured? Especially, because we install it either way right after this `if else`?

Reply:
I didn't find other ways to determine the PERMISSIVE vs STRICT in Pilot. What we have is the filter chains generated by the auth plugin.

I was using the number of filter chain to determine the PERMISSIVE vs STRICT. I think it's similar to this. The TLS inspector is also installed by auth plugins if it's PERMISSIVE mTLS. If the mTLS is PERMISSIVE mode, it will generate 2 filter chains with TLS inspector and if the mTLS is STRICT mode, it will generate 1 filter chain without TLS inspector.

Reply:
I vaguely remembered that by the time we introduced `PERMISSIVE`, it's the first time we introduce more than one filter chain. And the pilot filter plugin is not quite expressive, and end up with structure as this.